### ISSUE TITLE
Splat Map?

### ISSUE THREAD

**Issue by Thundros** [mebiusbox/pixy.js#1]
Hi, first off I must say you have done INCREDIBLE work with this project! I love everything you have done this far, please keep up the great work! Second, I'd like to ask if you could make a Terrain Texture Splat Map example, it would really benefit myself & many many other thousands of people. It would really help myself & others out a whole lot! Thank you SO much for everything you do & you have my thanks!

**Comment by mebiusbox** [comment]
@Thundros Thank you very much.

> Second, I'd like to ask if you could make a Terrain Texture Splat Map example, it would really benefit myself & many many other thousands of people. It would really help myself & others out a whole lot! 

That sounds interesting. Perhaps it is possible. The height map for a terrain generate from a noise, etc. but, I do not know if its quality of the texture is sufficient. Should I make something like a mountainous area?

**Comment by Thundros** [comment]
The quality would be phenomenal if you allow the shader the ability to change each texture size that is 'splatted' onto the map using a splat map to tell the Shader where on the terrain to place each texture. Also, the ability to use alpha textures is a MUST. This is the result of the ability for the Shader to change the size of each Texture : https://youtu.be/a05vtNcUYDU?t=1051 | Also, here is an example of a splat map with alpha & everything : https://blender.stackexchange.com/questions/49705/texture-splatting-mix-multiple-4-textures | Also, it would be advantageous to allow the mixing of more than 4 Textures at a time with RGBA as you can mix I hear, up to 256 Textures at once, see the Blender link for details. The Terrain heightmap MUST be seamless. A mountainous area with a green valley below & snow at the peaks would be a perfect example! <3 Because you get a little bit of each texture in one terrain!

**Comment by mebiusbox** [comment]
I made a terrain rendering using pixy.js in the past.

http://mebiusbox.github.io/contents/pixyjs/samples/shader_overlay.html

This demo is 1 base (gravel) and 4 layer (grass, cliff, snow, grassdark) texture blend using 4 channels of texture (splatmap). 

I might be wrong but... Do you want to make this texture by EffectTextureMaker?

**Comment by Thundros** [comment]
Whatever you think would produce more quality & still keep a 60 FPS frame rate

**Comment by Thundros** [comment]
Having any luck?

**Comment by mebiusbox** [comment]
It is better to create the height map from FBM noise. Snow can be created from a height map. Other layers can be created from the height map and some noise. I think that you can create a splatmap by creating a height map with EffectTextureMaker and processing it. However, quality is not very good. There are lots of tools to make splatmap, so you'd better try that. thanks.

**Comment by Thundros** [comment]
I simply think the ability to read a splatmap would be ideal because you can then paint stuff such as roads, trails, etc.

**Comment by Thundros** [comment]
If it's a splat map, you have full control over where all the roads, lakes, etc are placed. I don't want to MAKE a splatmap, I want to USE a splat map & be able to use a Shader to control the sizes of each texture{s}.

**Comment by mebiusbox** [comment]
here is splatmap example using pixy.js

https://github.com/mebiusbox/pixy.js/blob/master/samples/shader_overlay.html

first, create splatmap shader:

```
shader = new PIXY.Shader();
...
shader.enable("OVERLAY"); // enable splatmap
shader.enable("OVERLAYNORMAL"); // enable normalmap
```

second, setup splatmap and textures:

```
// layered textures
shader.uniforms.tOverlay1.value = loadTexture(textureLoader, 'assets/textures/Tiles/gravel_DM.png');
shader.uniforms.tOverlay2.value = loadTexture(textureLoader, 'assets/textures/Tiles/grass_DM.png');
shader.uniforms.tOverlay3.value = loadTexture(textureLoader, 'assets/textures/Tiles/cliff_DM.png');
shader.uniforms.tOverlay4.value = loadTexture(textureLoader, 'assets/textures/Tiles/snow_DM.png');
shader.uniforms.tOverlay5.value = loadTexture(textureLoader, 'assets/textures/Tiles/grassDark_DM.png');
// splatmap
shader.uniforms.tOverlayMask.value = loadTexture(textureLoader, 'assets/textures/Mask.png');
// layered texture scale
shader.uniforms.overlay1Scale.value = 1.0;
shader.uniforms.overlay2Scale.value = 1.0;
shader.uniforms.overlay3Scale.value = 1.0;
shader.uniforms.overlay4Scale.value = 1.0;
shader.uniforms.overlay5Scale.value = 1.0;
// layered normal textures
shader.uniforms.tOverlay1Normal.value = loadTexture(textureLoader, 'assets/textures/Tiles/gravel_NM.png');
shader.uniforms.tOverlay2Normal.value = loadTexture(textureLoader, 'assets/textures/Tiles/grass_NM.png');
shader.uniforms.tOverlay3Normal.value = loadTexture(textureLoader, 'assets/textures/Tiles/cliff_NM.png');
shader.uniforms.tOverlay4Normal.value = loadTexture(textureLoader, 'assets/textures/Tiles/Snow_NM.jpg');
shader.uniforms.tOverlay5Normal.value = loadTexture(textureLoader, 'assets/textures/Tiles/grass_NM.png');
```

You can use PIXY.HeightField to generate the terrain model. Please see the example for details.

**Comment by Thundros** [comment]
This is AMAZING! THANK YOU! <3 I just have one question. How exactly did you create that Splat map? How would I do it in Paint.net? I want to be able to paint my own roads.

**Comment by mebiusbox** [comment]
I got this texture from the internet. I think that any kind of software can be made if there is a brush function.

**Comment by Thundros** [comment]
So wait. You mean to tell me that I can paint ANY type of road I want with either Red, Green, Blue, or Alpha on the Splatmap itself & still have it paint that correct Texture onto the Heightmap with this shader?

**Comment by mebiusbox** [comment]
No. There is no function to paint splatmap directly in the example. This shader has a function to draw using pre-created splatmap. In order to paint you need to add some features.

**Comment by Thundros** [comment]
No, what I mean is, I can draw any type of road on the splatmap texture in paint.net & have it not mess up the textures on the landscape. i'll post a screen shot.

**Comment by Thundros** [comment]
My apologies @mebiusbox  for my absence, I have been busy working on my project. I will get a screenshot posted up within a month or 2. Thank you!

**Comment by mebiusbox** [comment]
This issue has been open for a long time without any updates. If you still have this problem or have additional information, please feel free to reopen the issue or create a new one. Thank you for your cooperation.
